File: src/main/java/com/ptah/auth/CustomUserDetailsService.java
package com.ptah.auth;

import com.ptah.common.Errors;
import com.ptah.repository.userprofiling.UserRepository;
import com.ptah.service.userprofiling.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class CustomUserDetailsService implements UserDetailsService {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private UserService userService;
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return Optional.ofNullable(userRepository.findByLogin(username)).map(user->CustomUserDetails.builder()
                .authorities(userService.getAuthorities(user))
                .password(user.getPassword())
                .build()).orElseThrow(()->new UsernameNotFoundException(Errors.INVALID_CREDENTIALS.getMessage()));
    }


}

File: backend/src/main/kotlin/com/gihub/jarryzhou/ptah/auth/CustomUserDetailsService.kt
package com.gihub.jarryzhou.ptah.auth

import com.ptah.common.Errors
import com.ptah.repository.userprofiling.UserRepository
import com.ptah.service.userprofiling.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import java.util.*

@Service
class CustomUserDetailsService : UserDetailsService {
    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var userService: UserService

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails = userRepository.findByLogin(username)?.let { user ->
        return CustomUserDetails().apply {
            authorities = userService.getAuthorities(user)
            internalPassword = user.password
            id = user.id
        }
    } ?: throw UsernameNotFoundException(Errors.INVALID_CREDENTIALS.message)
}
